### PR TITLE
Edge Shading

### PR DESCRIPTION
**Pull request by OscarSaharoy**
Hi, hope this project is still active :) I love jp2a and I had an idea which I think could be cool.

I added a few changes which allow the user to set an `--edge-threshold`. Anywhere that the image gradient is above this value, a directional character like `/|\=` is inserted, which can help highlight edges. I think the implementation could maybe be a bit better so any ideas appreciated!

Here is a demo, you can see the whiskers kind of get highlighted a bit:

```
$ ./src/jp2a --color --term-width --edge-threshold=0.3 ~/storage/pictures/Screenshots/Screenshot_20231126-174650_Gallery.png
```

IMG_20231128_200651

IMG_20231128_203003

### CODE REVIEW

File: src/image.c [hunk 1]
```diff
@@ -283,53 +324,46 @@ void print_image_no_colors(const Image* const image, const int chars, FILE *f) {
...
+			size_t char_len = ascii_palette_lengths[i];
 #endif
-		}
-#if ASCII
-		fprintf(f, !use_border? "%s\n" : "|%s|\n", line);
-#else
-		if ( !flipx ) {
+			const vec2 gradient = get_image_gradient(image, x, y);
+			if( i > 1 && magnitude(gradient) > edge_threshold ) {
+				float direction_scaled = direction(gradient) / M_PI * 4. + 5.5;
```
Comment (Talinx) [inline src/image.c:353]: Where does 5.5 come from?
I guess this is 360° - 45° = 2 * π - (π/4). Rewriting this using π or a comment would be nice.

File: src/image.c [hunk 2]
```diff
@@ -79,51 +82,91 @@ void print_image(Image *image, FILE *f) {
...
+#if ASCII
+			char* char_start = &ascii_palette[i];
+			size_t char_len = 1;
+#else
+			char* char_start = &ascii_palette[ascii_palette_indizes[i]];
+			size_t char_len = ascii_palette_lengths[i];
 #endif
+			const vec2 gradient = get_image_gradient(image, x, y);
+			if( i > 1 && magnitude(gradient) > edge_threshold ) {
+				float direction_scaled = direction(gradient) / M_PI * 4. + 5.5;
```
Comment (Talinx) [inline src/image.c:163]: (See other comment.)

**Review by Talinx — CHANGES_REQUESTED** [review]
Great idea and implementation!

I got some more ideas (if you want to implement more):
- An option to suppress non-edge output would be neat, i.e. like an edge-detector.
- Support for different common edge-detection kernels could also be added.

(Also one test (`make test`, test ( 9) (width, border)) now fails. If you want you can look into this, otherwise I'll fix that.)

File: src/image.c (hunk 1 above)
Comment (OscarSaharoy) [reply src/image.c]: Yes I think so, there is a +4 there that is needed to make the `fmod` call always return a positive number too. I have rewritten this part and added a comment :)

File: src/image.c [hunk 3]
```diff
@@ -79,51 +82,91 @@ void print_image(Image *image, FILE *f) {
...
+vec2 get_image_gradient(const Image* const image, const int x, const int y) {
+	const float kernel_x[9] = {
+		-4./4.,    0.,
+		 4./4., 0./4.
+	};
+	const float kernel_y[9] = {
+		-4./4., 4./4.,
+		    0., 0./4.
+	};
+	const float patch[9] = {
```
[the remark this answers is not in the document]
Comment (OscarSaharoy) [reply src/image.c]: Thanks for spotting that!

**Comment by OscarSaharoy** [comment]
Thanks so much for taking a look :) the test failure was due to accessing outside the image array but should be fixed now. And thanks for the edges only suggestion! it works really nice:

Screenshot from 2023-12-10 13-39-39

For the alternative edge detection kernels, I did try a couple of them but the simple one I have used seemed the best. I think the subsampling of the initial image means that the edge detection is already kind of awkward as edges are blurred out a bit, so to take advantage of different edge kernels it would probably make more sense to apply to kernel before subsampling. Maybe that can be a future feature :) The test script is also very cool, I can add some tests for edge shading if you like? But I guess you might have a better idea of what to add than me.

**Comment by Talinx** [comment]
Makes sense that other kernels do not look really different for ASCII art.

Adding three little test (one with and one without `--edges-only` and one with both `--edges-only` and `--invert` would be nice, using `tests/jp2a.jpg` sounds good to me.

Two little things are still missing:

- bash completion for the new command options (`completion/bash/jp2a`)
- adding the new command options to the man page (`man/jp2a.1`)

Also when using `--edges-only` and `--invert`, empty space/spaces are replaced with `M`. Is this intended? (This happens because `pos` is set to 0 (due to `--edges-only`), which then selects the last char from the ASCII chars (`M`) because the first pos is switched with the last (due to `--invert`).)

This is coming together very nicely ;)

**Comment by OscarSaharoy** [comment]
The man page and autocomplete should be updated now :) and I also added a commit to prevent a space being added when you autocomplete an option that ends in =

Before I do the tests I am just wondering about the behaviour with `--edges-only` and `--invert` - I think the current behaviour is ok because it gives the user the option to choose whether the first or last character in the palette is used to fill the rest if the image, which seem like the two logical choices. With certain custom palettes this could be useful? If there is a better way to handle it though I will change it and then make the tests :)

**Comment by Talinx** [comment]
The current behavior for `--invert` looks good to me. Just wanted to make sure this code path is thought about.

**Comment by OscarSaharoy** [comment]
Ok perfect, the tests should be there now :)

**Review by Talinx — APPROVED** [review]

**Comment by Talinx** [comment]
Awesome! Unless you want to add something else to this PR it's ready to merge.

**Comment by OscarSaharoy** [comment]
Thanks so much! Sounds good to me, I really enjoyed working on this and learned a lot :)
